tools/payments/cmd/prepare/main.go

```diff
@@ -85,41 +104,99 @@ func main() {
 		log.Fatal("failed payout id cannot be nil or empty\n")
 	}
 
-	wc := &payments.WorkerConfig{
-		PayoutID:      *payoutID,
-		ConsumerGroup: payments.PrepareStream + "-cg",
-		Stream:        payments.PrepareStream,
-		Count:         0,
+	firstRun := true
+	// FIXME
+	responseStream := payments.PreparePrefix + *payoutID + payments.ResponseSuffix
+	responseFile := responseStream + ".log"
+	if _, err := os.Stat(responseFile); err == nil {
+		firstRun = false
+		if !*resubmit {
+			log.Println("not first run, skipping add to prepare stream")
+		}
 	}
 
-	for _, name := range files {
-		func() {
-			f, err := os.Open(name)
-			if err != nil {
-				log.Fatalf("failed to open report file: %v\n", err)
-			}
-			defer f.Close()
+	totalTransactions := 0
+	f, err := os.Open(files[0])
+	if err != nil {
+		log.Fatalf("failed to open report file: %v\n", err)
+	}
+	defer f.Close()
 
-			report := payments.PreparedReport{}
-			if err := payments.ReadReport(&report, f); err != nil {
-				log.Fatalf("failed to read report from stdin: %v\n", err)
-			}
+	report := paymentscli.PreparedReport{}
+	if err := paymentscli.ReadReport(&report, f); err != nil {
+		log.Fatalf("failed to read report from stdin: %v\n", err)
+	}
+	if err := report.Validate(); err != nil {
+		log.Fatalf("failed to validate report: %v\n", err)
+	}
 
-			wc.Count += len(report)
+	if report[0].PayoutID != *payoutID {
+		log.Fatalf("payoutID did not match report: %s\n", report[0].PayoutID)
+	}
 
-			if err := report.Prepare(ctx, client); err != nil {
-				log.Fatalf("failed to read report from stdin: %v\n", err)
-			}
-		}()
+	totalTransactions += len(report)
+	if firstRun || *resubmit {
+		priv, err := paymentscli.GetOperatorPrivateKey(*key)
+		if err != nil {
+			log.Fatalf("failed to parse operator key file: %v\n", err)
+		}
+
+		if err := report.Prepare(ctx, priv, client); err != nil {
+			log.Fatalf("failed to read report from stdin: %v\n", err)
+		}
+
+		wc := &payments.WorkerConfig{
+			PayoutID:      *payoutID,
+			ConsumerGroup: payments.PreparePrefix + *payoutID + "-cg",
+			Stream:        payments.PreparePrefix + *payoutID,
+			Count:         len(report),
+		}
+
+		err = client.ConfigureWorker(ctx, payments.PrepareConfigStream, wc)
+		if err != nil {
+			log.Fatalf("failed to write to prepare config stream: %v\n", err)
+		}
+		if *verbose {
+			log.Printf("prepare transactions loaded for %+v\n", payoutID)
+		}
+
+		os.Create(responseFile)
 	}
 
-	err = client.ConfigureWorker(ctx, payments.PrepareConfigStream, wc)
+	// FIXME default to public key as consumer group?
+	err = client.WaitForResponses(ctx, *payoutID, totalTransactions, responseStream, *cg)
```

Comment:
I think this is causing me to get the prepare step to hang currently. Can we double check the logic here. I'm seeing an error where the command runs properly but it sits waiting for a response the whole time.

Reply:
https://github.com/brave-intl/bat-go/issues/2385